src/peft/tuners/lora/layer.py

```diff
@@ -232,6 +236,53 @@ def unscale_layer(self, scale=None) -> None:
                 self.scaling[active_adapter] /= scale
 
 
+def onload_delta_wrapper(func):
+    """
+    Performs LoRA parameter onloading and offloading required for get_delta_weight() to function properly with
+    offloaded modules.
+    """
+
+    def wrapper(self, adapter):
+        if hasattr(self.lora_A[adapter], "_hf_hook") and isinstance(self.lora_A[adapter]._hf_hook, AlignDevicesHook):
+            self.lora_A[adapter]._hf_hook.pre_forward(self.lora_A[adapter])
+        if hasattr(self.lora_B[adapter], "_hf_hook") and isinstance(self.lora_B[adapter]._hf_hook, AlignDevicesHook):
+            self.lora_B[adapter]._hf_hook.pre_forward(self.lora_B[adapter])
+
+        output = func(self, adapter)
+
+        if hasattr(self.lora_A[adapter], "_hf_hook") and isinstance(self.lora_A[adapter]._hf_hook, AlignDevicesHook):
+            self.lora_A[adapter]._hf_hook.post_forward(self.lora_A[adapter], torch.tensor([]))
+        if hasattr(self.lora_B[adapter], "_hf_hook") and isinstance(self.lora_B[adapter]._hf_hook, AlignDevicesHook):
+            self.lora_B[adapter]._hf_hook.post_forward(self.lora_B[adapter], torch.tensor([]))
+        return output
+
+    return wrapper
+
+
+@contextmanager
+def onload_layer(base_layer):
+    if (
+        hasattr(base_layer, "_hf_hook")
+        and isinstance(base_layer._hf_hook, AlignDevicesHook)
+        and base_layer._hf_hook.offload
+    ):
+        base_layer._hf_hook.pre_forward(base_layer)
+
+    yield
+
+    if (
+        hasattr(base_layer, "_hf_hook")
+        and isinstance(base_layer._hf_hook, AlignDevicesHook)
+        and base_layer._hf_hook.offload
+    ):
+        base_layer._hf_hook.weights_map = {name: param.to("cpu") for name, param in named_module_tensors(base_layer)}
+        # offload weights map to disk if original device is the disk
+        if torch.device("meta") in base_layer._hf_hook.original_devices.values():
+            offload_folder = tempfile.TemporaryDirectory()
```

Comment:
Can we use the `TemporaryDirectory` as a context manager, otherwise I think there is no guarantee that it will be cleaned up. Or is it necessary for it _not_ to be cleaned up right away?

Reply:
On second thought I think it is indeed best to have a string for the directory here, as this is how the disk-offloaded directory is usually initially specified.  We do need the directory to be persistent, as the updated disk-offloaded weights are saved here. The original method (making a new folder) was I suppose not very efficient, so the updated way is to find the original offload directory and update that.